File: calc.cpp
#include <node.h>
#include <v8.h>
#include <stdio.h>
#include <stdlib.h>

using namespace v8;

Handle<Value> getSquare(const Arguments& args) {
    HandleScope scope;
    
    if(args.Length() < 1) {
        ThrowException(Exception::TypeError(String::New("Wrong number of arguments")));
        return scope.Close(Undefined());
    }
    
    Local<Number> x = args[0]->ToNumber();

    Local<Object> obj = Object::New();
    obj->Set(String::NewSymbol("x"), Number::New( x->IntegerValue() * x->IntegerValue()));
    return scope.Close(obj);
}

void init(Handle<Object> target) {
    target->Set(String::NewSymbol("getSquare")
    , FunctionTemplate::New(getSquare)->GetFunction());
}

NODE_MODULE(calc, init);

File: calc.js
var calc = require('./build/Release/calc.node');

var x = 1;
console.log(x + ' in square: ' + calc.getSquare(x).x);
var x = 2;
console.log(x + ' in square: ' + calc.getSquare(x).x);
var x = 3;
console.log(x + ' in square: ' + calc.getSquare(x).x);
var x = 4;
console.log(x + ' in square: ' + calc.getSquare(x).x);

console.log('no args: ' + calc.getSquare().x);
